File: src/main/java/com/learning/leetcode/recursive/easy/FloodFill.py
from typing import List


def floodFill(image: List[List[int]], sr: int, sc: int, newColor: int):
    floodFillHelper(image, sr, sc, image[sr][sc], newColor)


def floodFillHelper(image: List[List[int]], i: int, j: int, oldColor: int, newColor: int):
    if i < 0 or j < 0 or i == len(image) or j == len(image[0]) or image[i][j] != oldColor or image[i][j] == newColor:
        return
    image[i][j] = newColor
    floodFillHelper(image, i + 1, j, oldColor, newColor)
    floodFillHelper(image, i - 1, j, oldColor, newColor)
    floodFillHelper(image, i, j + 1, oldColor, newColor)
    floodFillHelper(image, i, j - 1, oldColor, newColor)


def main():
    image = [[1, 1, 1], [1, 1, 0], [1, 0, 1]]
    sr = 1
    sc = 1
    newColor = 2
    print("Given image")
    for row in image:
        print(row)
    print("Selected color", image[sr][sc])
    print("New color", newColor)
    floodFill(image, sr, sc, newColor)
    print("New image")
    for row in image:
        print(row)



main()

File: src/main/java/com/learning/leetcode/recursive/easy/FloodFill.java
package com.learning.leetcode.recursive.easy;

import java.util.Arrays;

public class FloodFill {

    public static void main(String[] args) {
        int[][] image = {{1,1,1},{1,1,0},{1,0,1}};
        int startingRow = 1;
        int startingColumn = 1;
        int newColor = 2;
        System.out.println("Initial image");
        Arrays.stream(image)
                .forEach(row -> System.out.println(Arrays.toString(row)));
        System.out.println("Starting point: " + startingRow + ", " + startingColumn);
        int oldColor = image[startingRow][startingColumn];
        flood(0, 0, image, oldColor, newColor);
        System.out.println("Post flood fill image");
        Arrays.stream(image)
                .forEach(row -> System.out.println(Arrays.toString(row)));
    }

    private static void flood(int i, int j, int[][] image, int oldColor, int newColor) {
        if (i < 0 || j < 0 || i == image.length || j == image[0].length  || image[i][j] != oldColor) {
            return;
        }
        image[i][j] = newColor;
        flood(i + 1, j, image, oldColor, newColor);
        flood(i - 1, j, image, oldColor, newColor);
        flood(i, j + 1, image, oldColor, newColor);
        flood(i, j - 1, image, oldColor, newColor);
    }

}
